cmd/manager/main.go

```diff
@@ -39,35 +49,71 @@ func printVersion(log logr.Logger) {
 	log.Info(fmt.Sprintf("Version of operator-sdk: %v", sdkVersion.Version))
 }
 
-// function to handle defers with exit, see https://stackoverflow.com/a/27629493/553720.
-func doMain() int {
-	pflag.CommandLine.AddFlagSet(zap.FlagSet())
-	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
-	pflag.Parse()
-
-	ctrl.SetLogger(zap.Logger())
-	log := ctrl.Log.WithName("cmd")
-
-	printVersion(log)
+// function to choose transport type based on env var.
+func getTransport(transportType string) (transport.Transport, error) {
+	switch transportType {
+	case kafkaTransportTypeName:
+		kafkaProducer, err := kafkaClient.NewHOHProducer(ctrl.Log.WithName("kafka-client"))
+		if err != nil {
+			return nil, fmt.Errorf("failed to create hoh-kafka-producer: %w", err)
+		}
+
+		return kafkaProducer, nil
+	case syncServiceTransportTypeName:
+		syncService, err := hohSyncService.NewSyncService(ctrl.Log.WithName("sync-service"))
+		if err != nil {
+			return nil, fmt.Errorf("failed to create sync-service: %w", err)
+		}
+
+		return syncService, nil
+	default:
+		return nil, errEnvVarIllegalValue
+	}
+}
 
+func readEnvVars(log logr.Logger) (string, time.Duration, string, error) {
 	leaderElectionNamespace, found := os.LookupEnv(envVarControllerNamespace)
 	if !found {
 		log.Error(nil, "Not found:", "environment variable", envVarControllerNamespace)
-		return 1
+		return "", 0, "", errEnvVarNotFound
```

Comment:
in all returning fields, in the places where we have error, readEnvVars shouldn't log the error.
it should just return wrapped error using fmt.Errorf(... %w, err).
main function should log the error and exit.

so no need to give this function the logger as an argument.
see for example here:
https://github.com/open-cluster-management/hub-of-hubs-spec-transport-bridge/blob/main/pkg/transport/sync-service/sync_service.go#L53